File: 16. Maximum area of rectangle in a binary matrix using stack.py
"""
Task: Maximum area of rectangle in a binary matrix using stack

Time Complexity: O(n*m)
Space Complexity: O(m)
"""

def largest_rectangle_area(heights):
    n = len(heights)
    st = []
    maxA = 0
    for i in range(n+1):
        h = 0 if i==n else heights[i]
        while st and h < heights[st[-1]]:
            height = heights[st.pop()]
            width = i if not st else i - st[-1] - 1
            maxA = max(maxA, height*width)
        st.append(i)
    return maxA


def max_area_in_binary_matrix(mat):
    if not mat: return 0
    n = len(mat); m = len(mat[0])
    heights = [0]*m
    best = 0
    for i in range(n):
        for j in range(m):
            heights[j] = heights[j]+1 if mat[i][j] else 0
        best = max(best, largest_rectangle_area(heights))
    return best

if __name__ == '__main__':
    mat = [[1,0,1,0,0],[1,0,1,1,1],[1,1,1,1,1],[1,0,0,1,0]]
    print(max_area_in_binary_matrix(mat))

File: 16. Maximum area of rectangle in a binary matrix using stack.cpp
/*
 Task: Maximum area of rectangle in a binary matrix using stack
*/

#include <bits/stdc++.h>
using namespace std;

int maxAreaInBinaryMatrix(vector<vector<int>>& mat)
{
    if (mat.empty())
        return 0;
    int n = mat.size(), m = mat[0].size();
    vector<int> heights(m, 0);
    int best = 0;
    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < m; ++j) heights[j] = mat[i][j] ? heights[j] + 1 : 0;
        best = max(best, [ & ](const vector<int>&h){ // compute largest rectangle in histogram
            int res=0; stack<int> st; int sz=h.size();
            for(int k=0;k<=sz;++k){ int hh = (k==sz)?0:h[k]; while(!st.empty() && hh < h[st.top()]){ int height=h[st.top()]; st.pop(); int width = st.empty()?k:k-st.top()-1; res=max(res,height*width);} st.push(k);} return res; }(heights));
    }
    return best;
}

int main()
{
    vector<vector<int>> mat = {{1, 0, 1, 0, 0}, {1, 0, 1, 1, 1}, {1, 1, 1, 1, 1}, {1, 0, 0, 1, 0}};
    cout << maxAreaInBinaryMatrix(mat) << "\n";
    return 0;
}

// Time Complexity: O(n*m)
// Space Complexity: O(m)
